apps/omg/lib/omg/state/core.ex

```diff
@@ -388,6 +432,49 @@ defmodule OMG.State.Core do
     }
   end
 
+  defp collect_fees(%Core{fees_paid: fees_paid} = state, token_surplussed) do
+    fees_paid_with_new =
+      token_surplussed
+      |> Enum.reject(fn {_token, amount} -> amount == 0 end)
+      |> Map.new()
+
+    %Core{
+      state
+      | fees_paid:
+          Map.merge(fees_paid, fees_paid_with_new, fn
+            _token, collected, tx_surplus -> collected + tx_surplus
+          end)
+    }
+  end
+
+  defp disallow_payments(state), do: %Core{state | fee_claiming_started: true}
+
+  defp flush_collected_fees_for_token(state, %Output{currency: token}),
+    do: %Core{state | fees_paid: Map.delete(state.fees_paid, token)}
+
+  @spec claim_fees(state :: t()) :: t()
+  defp claim_fees(
+         %Core{
+           height: height,
+           fees_paid: fees_paid,
+           fee_claimer_address: owner
+         } = state
+       ) do
+    new_state =
+      height
+      |> Transaction.FeeTokenClaim.claim_collected(owner, fees_paid)
+      |> Enum.map(fn fee_tx ->
+        Transaction.Signed.encode(%Transaction.Signed{raw_tx: fee_tx, sigs: []})
+      end)
+      |> Enum.reduce(state, fn rlp_tx, curr_state ->
+        {:ok, tx} = Transaction.Recovered.recover_from(rlp_tx)
+        {:ok, _, new_state} = exec(curr_state, tx, :no_fees_required)
+        new_state
+      end)
+
+    %Core{new_state | fees_paid: %{}, fee_claiming_started: false}
```

Comment:
also `fees_paid: %{}` seems unnecessary at 1st glance - doesn't the `flush` function do that already?